projects/media-viewer/src/lib/media-viewer/viewers/pdf-viewer/pdf-viewer.component.ts

```diff
@@ -43,12 +43,15 @@ export class PdfViewerComponent implements AfterViewInit, OnDestroy {
   }
 
   handleMessage(genericMessage: MediaViewerMessage) {
+    if (genericMessage instanceof RotateSinglePageOperation) {
```

Comment:
This method looks like a good candidate for https://refactoring.com/catalog/replaceConditionalWithPolymorphism.html

My suggestion would be to have services that subscribe to particular types of message (rotate, zoom, etc) and then emit the event based on the message type. I believe the angular way is to use RX.js observables but @benouaer knows better than I do

Reply:
there is an angular library [https://ngrx.io/guide/store](ngrx) that I've mentioned to pawel that handles this kind of behaviour using 'Actions' of different types that you dispatch in different scenarios, and which are intercepted by different 'Effects' that can perform the steps you need to do.
there's some boilerplate that comes with using this library, but I think it would be worth it since there are quite a few bits of state to manage. the good thing with this is that the components are very clean, as all the state management is moved to these specialised classes.

obviously, we can always replicate the same approach ourselves if we don't have that many actions to manage

Reply:
actually, now having looked at the code a bit more closely, it seems to me that our state is contained mainly within 'pdfViewer'. so regardless of whether or not we use ngrx, if we maintained that within a store object, and then exposed functions that allow us to directly either select or update from it, that would remove the need for messages.